ethereum-side/NFTLock.sol

```diff
@@ -12,173 +12,97 @@ contract NFTLock is INFTLock, IERC721Receiver {
-
-        token.state == State.unlocked;
+        releasedIds.push(internalTokenId);
+        delete lockedTokenById[internalTokenId];
+        totalDepositedTokens--;
 
         emit TokenUnlocked(internalTokenId, msg.sender);
     }
 
-    function withdraw(uint256 internalTokenId) external override isDeposited(internalTokenId) {
-
-        NFT storage token = tokens[msg.sender][internalTokenId];
+    function withdraw(uint256 internalTokenId) external override {
 
-        require(token.state == State.unlocked, "NFT Locker: Unlock you token before to withdraw");
+        NFT storage token = lockedTokenById[internalTokenId];
+        address sendTo = token.depositor;
 
-        ERC721(token.tokenContractAddress).safeTransferFrom(address(this), msg.sender, token.externalTokenId);
+        if (isMintedOnTz(internalTokenId)) {
 
-        uint256 totalTokens = totalTokensByAddress[msg.sender];
+            address newNFTOwner;
+            bool isBurned;
 
-        // switch NFT positions in order to optimize a gas
-        if (internalTokenId != totalTokens - 1) {
-            token = tokens[msg.sender][totalTokens - 1];
+            (isBurned, newNFTOwner) = isBurnedOnTz(internalTokenId);
+            require(isBurned, "NFT Locker: Burn your wrapped token before to unlock");
+            sendTo = newNFTOwner;
         }
-        totalTokensByAddress[msg.sender]--;
-        totalDepositedTokens--;
 
-        emit NFTWithdrawn(internalTokenId, msg.sender);
-    }
+        uint256 externalTokenId = token.externalTokenId;
+        ERC721(token.tokenContractAddress).safeTransferFrom(address(this), sendTo, externalTokenId);
 
-    function tokenWrapped(address NFTOwner, uint256 internalTokenId) external override onlyAdmin {
+        // delete NFT from the storage
+        unlockToken(internalTokenId);
 
-        NFT storage token = tokens[NFTOwner][internalTokenId];
-        token.state = State.wrapMinted;
+        emit NFTWithdrawn(internalTokenId, sendTo);
     }
 
-    function tokenBurned(address oldNFTOwner, address newNFTOwner, uint256 internalTokenId) external override onlyAdmin {
-
-
-        NFT storage token = tokens[oldNFTOwner][internalTokenId];
-
-        if (oldNFTOwner != newNFTOwner) {
-
-            uint256 totalTokens = totalTokensByAddress[oldNFTOwner];
-
-            // remove token from an old owner
-            if (internalTokenId != totalTokens - 1) {
-                token = tokens[oldNFTOwner][totalTokens - 1];
-            }
-            totalTokensByAddress[oldNFTOwner]--;
-
-            // add token to a new owner
-            totalTokens = totalTokensByAddress[newNFTOwner];
-            tokens[newNFTOwner][totalTokens] = token;
-            totalTokensByAddress[newNFTOwner]++;
-        }
-
-        token.state = State.wrapBurned;
+    //it's a mock of a call to oracle function
+    function isMintedOnTz(uint256 internalTokenId) internal pure returns (bool) {
+        internalTokenId++;
```

Comment:
`internalTokenId++` ? I don't understand ...

Reply:
Same in `isBurned`

Reply:
It doesn't make sense actually, just to remove compilator warnings about an unused variable^^